File: editor/pane.go
package editor

// Pane represents a visual view of a buffer with independent navigation state.
// Multiple panes can reference the same buffer (shared buffer model).
type Pane struct {
	Buffer *Buffer // Reference to shared buffer

	// Independent cursor position (each pane has its own cursor)
	CursorRow int
	CursorCol int

	// Independent scroll position
	ScrollOffset int // First line displayed at top of screen

	// Independent selection state for Visual mode
	SelectionActive   bool
	SelectionStartRow int
	SelectionStartCol int
}

// NewPane creates a new pane viewing the given buffer
func NewPane(buf *Buffer) *Pane {
	return &Pane{
		Buffer:    buf,
		CursorRow: 0,
		CursorCol: 0,
	}
}

// NewPaneAtLine creates a new pane viewing the given buffer, positioned at a specific line
func NewPaneAtLine(buf *Buffer, line int) *Pane {
	p := NewPane(buf)
	if line > 0 {
		p.GotoLine(line)
	}
	return p
}

// GotoLine moves the cursor to a specific line (1-based for user, converts to 0-based)
func (p *Pane) GotoLine(line int) {
	if line < 1 {
		line = 1
	}
	if line > len(p.Buffer.Lines) {
		line = len(p.Buffer.Lines)
	}
	p.CursorRow = line - 1
	p.CursorCol = 0
	p.clampCursor()
}

// clampCursor ensures the cursor position is within valid bounds
func (p *Pane) clampCursor() {
	// Clamp row to buffer bounds
	if p.CursorRow < 0 {
		p.CursorRow = 0
	}
	if p.CursorRow >= len(p.Buffer.Lines) {
		p.CursorRow = len(p.Buffer.Lines) - 1
	}
	if p.CursorRow < 0 {
		p.CursorRow = 0 // Handle empty buffer edge case
	}

	// Clamp column to line length
	p.clampCursorCol()
}

// clampCursorCol ensures the cursor column is within the current line bounds
func (p *Pane) clampCursorCol() {
	if len(p.Buffer.Lines) == 0 {
		p.CursorCol = 0
		return
	}
	lineLen := len(p.Buffer.Lines[p.CursorRow])
	if p.CursorCol > lineLen {
		p.CursorCol = lineLen
	}
	if p.CursorCol < 0 {
		p.CursorCol = 0
	}
}

// AdjustCursorForEdit adjusts the cursor position when lines are added or removed
// editRow is the row where the edit occurred
// linesDelta is positive for insertions, negative for deletions
func (p *Pane) AdjustCursorForEdit(editRow, linesDelta int) {
	if linesDelta == 0 {
		return
	}

	if linesDelta > 0 {
		// Lines inserted: shift cursor down if it's at or after the edit
		if p.CursorRow >= editRow {
			p.CursorRow += linesDelta
		}
	} else {
		// Lines deleted
		deletedStart := editRow
		deletedEnd := editRow + (-linesDelta) - 1

		if p.CursorRow > deletedEnd {
			// Cursor is after deleted range: shift up
			p.CursorRow += linesDelta
		} else if p.CursorRow >= deletedStart {
			// Cursor is within deleted range: move to deletion point
			p.CursorRow = deletedStart
		}
		// If cursor is before deletion, no change needed
	}

	// Always clamp after adjustment
	p.clampCursor()
}

// AdjustScroll adjusts the scroll offset to keep the cursor visible
// with a margin from the top and bottom edges
// textWidth is the available width for text (excluding line numbers)
// screenHeight is the number of visible lines
func (p *Pane) AdjustScroll(textWidth, screenHeight int) {
	if textWidth < 1 {
		textWidth = 1
	}

	// Scroll margin - lines to keep visible above/below cursor
	const scrollMargin = ScrollMargin

	// Adjust margin if screen is too small
	margin := scrollMargin
	if screenHeight < margin*2+1 {
		margin = (screenHeight - 1) / 2
	}
	if margin < 0 {
		margin = 0
	}

	// Calculate which wrapped row the cursor is on within its line
	cursorWrapRow := 0
	if p.CursorCol > 0 && textWidth > 0 {
		visualCol := p.Buffer.GetVisualColumn(p.Buffer.Lines[p.CursorRow], p.CursorCol)
		cursorWrapRow = visualCol / textWidth
	}

	// Calculate screen rows from ScrollOffset to cursor (top margin check)
	screenRowsFromTop := 0
	for lineIdx := p.ScrollOffset; lineIdx < p.CursorRow && lineIdx < len(p.Buffer.Lines); lineIdx++ {
		visualWidth := p.Buffer.GetVisualLineWidth(p.Buffer.Lines[lineIdx])
		if visualWidth == 0 {
			screenRowsFromTop++
		} else {
			screenRowsFromTop += (visualWidth + textWidth - 1) / textWidth
		}
	}
	screenRowsFromTop += cursorWrapRow

	// If cursor is too close to top, scroll up
	if screenRowsFromTop < margin && p.ScrollOffset > 0 {
		for screenRowsFromTop < margin && p.ScrollOffset > 0 {
			p.ScrollOffset--
			visualWidth := p.Buffer.GetVisualLineWidth(p.Buffer.Lines[p.ScrollOffset])
			if visualWidth == 0 {
				screenRowsFromTop++
			} else {
				screenRowsFromTop += (visualWidth + textWidth - 1) / textWidth
			}
		}
	}

	// Calculate total screen rows used including cursor position
	screenRowsUsed := screenRowsFromTop + 1

	// If cursor is too close to bottom, scroll down
	if screenRowsUsed > screenHeight-margin {
		excess := screenRowsUsed - (screenHeight - margin)
		for excess > 0 && p.ScrollOffset < p.CursorRow {
			visualWidth := p.Buffer.GetVisualLineWidth(p.Buffer.Lines[p.ScrollOffset])
			lineRows := 1
			if visualWidth > 0 {
				lineRows = (visualWidth + textWidth - 1) / textWidth
			}
			if lineRows <= excess {
				excess -= lineRows
				p.ScrollOffset++
			} else {
				break
			}
		}
		if excess > 0 {
			p.ScrollOffset++
		}
	}

	// Ensure cursor is never above ScrollOffset
	if p.CursorRow < p.ScrollOffset {
		p.ScrollOffset = p.CursorRow - margin
		if p.ScrollOffset < 0 {
			p.ScrollOffset = 0
		}
	}
}

// StartSelection begins a visual selection at the current cursor position
func (p *Pane) StartSelection() {
	p.SelectionActive = true
	p.SelectionStartRow = p.CursorRow
	p.SelectionStartCol = p.CursorCol
}

// ClearSelection ends the visual selection
func (p *Pane) ClearSelection() {
	p.SelectionActive = false
}

// GetSelection returns the selection bounds normalized (start before end)
func (p *Pane) GetSelection() (startRow, startCol, endRow, endCol int) {
	return NormalizeRange(p.SelectionStartRow, p.SelectionStartCol, p.CursorRow, p.CursorCol)
}

// IsInSelection returns true if the given position is within the selection
func (p *Pane) IsInSelection(row, col int) bool {
	if !p.SelectionActive {
		return false
	}

	startRow, startCol, endRow, endCol := p.GetSelection()

	// Before start
	if row < startRow || (row == startRow && col < startCol) {
		return false
	}

	// After end
	if row > endRow || (row == endRow && col > endCol) {
		return false
	}

	return true
}

// Movement methods that delegate to buffer but update pane's cursor

// MoveUp moves the cursor up one line
func (p *Pane) MoveUp() {
	if p.CursorRow > 0 {
		p.CursorRow--
		p.clampCursorCol()
	}
}

// MoveDown moves the cursor down one line
func (p *Pane) MoveDown() {
	if p.CursorRow < len(p.Buffer.Lines)-1 {
		p.CursorRow++
		p.clampCursorCol()
	}
}

// MoveLeft moves the cursor left one character
func (p *Pane) MoveLeft() {
	if p.CursorCol > 0 {
		p.CursorCol--
	} else if p.CursorRow > 0 {
		p.CursorRow--
		p.CursorCol = len(p.Buffer.Lines[p.CursorRow])
	}
}

// MoveRight moves the cursor right one character
func (p *Pane) MoveRight() {
	if p.CursorCol < len(p.Buffer.Lines[p.CursorRow]) {
		p.CursorCol++
	} else if p.CursorRow < len(p.Buffer.Lines)-1 {
		p.CursorRow++
		p.CursorCol = 0
	}
}

// MoveUpN moves the cursor up N lines
func (p *Pane) MoveUpN(n int) {
	for i := 0; i < n && p.CursorRow > 0; i++ {
		p.CursorRow--
	}
	p.clampCursorCol()
}

// MoveDownN moves the cursor down N lines
func (p *Pane) MoveDownN(n int) {
	for i := 0; i < n && p.CursorRow < len(p.Buffer.Lines)-1; i++ {
		p.CursorRow++
	}
	p.clampCursorCol()
}

// MoveLeftN moves the cursor left N characters
func (p *Pane) MoveLeftN(n int) {
	for i := 0; i < n; i++ {
		p.MoveLeft()
	}
}

// MoveRightN moves the cursor right N characters
func (p *Pane) MoveRightN(n int) {
	for i := 0; i < n; i++ {
		p.MoveRight()
	}
}

// MoveToLineStart moves the cursor to the beginning of the line
func (p *Pane) MoveToLineStart() {
	p.CursorCol = 0
}

// MoveToLineEnd moves the cursor to the end of the line
func (p *Pane) MoveToLineEnd() {
	p.CursorCol = len(p.Buffer.Lines[p.CursorRow])
}

// PageDown moves the cursor down by height lines
func (p *Pane) PageDown(height int) {
	p.CursorRow += height / 2
	if p.CursorRow >= len(p.Buffer.Lines) {
		p.CursorRow = len(p.Buffer.Lines) - 1
	}
	p.clampCursorCol()
}

// PageUp moves the cursor up by height lines
func (p *Pane) PageUp(height int) {
	p.CursorRow -= height / 2
	if p.CursorRow < 0 {
		p.CursorRow = 0
	}
	p.clampCursorCol()
}

// MoveToNextWord moves the cursor to the start of the next word
func (p *Pane) MoveToNextWord() {
	line := p.Buffer.Lines[p.CursorRow]

	if p.CursorCol < len(line) {
		ch := line[p.CursorCol]
		if isWordChar(ch) {
			// Skip word characters
			for p.CursorCol < len(line) && isWordChar(line[p.CursorCol]) {
				p.CursorCol++
			}
		} else if !isWhitespace(ch) {
			// Skip punctuation characters
			for p.CursorCol < len(line) && !isWordChar(line[p.CursorCol]) && !isWhitespace(line[p.CursorCol]) {
				p.CursorCol++
			}
		}
	}

	// Skip whitespace
	for p.CursorCol < len(line) && isWhitespace(line[p.CursorCol]) {
		p.CursorCol++
	}

	// If at end of line, move to start of next line
	if p.CursorCol >= len(line) && p.CursorRow < len(p.Buffer.Lines)-1 {
		p.CursorRow++
		p.CursorCol = 0
		// Skip leading whitespace on new line
		line = p.Buffer.Lines[p.CursorRow]
		for p.CursorCol < len(line) && isWhitespace(line[p.CursorCol]) {
			p.CursorCol++
		}
	}
}

// MoveToPrevWord moves the cursor to the start of the previous word
func (p *Pane) MoveToPrevWord() {
	// If at start of line, move to end of previous line
	if p.CursorCol == 0 && p.CursorRow > 0 {
		p.CursorRow--
		p.CursorCol = len(p.Buffer.Lines[p.CursorRow])
	}

	line := p.Buffer.Lines[p.CursorRow]

	// Move back one if not at start
	if p.CursorCol > 0 {
		p.CursorCol--
	}

	// Skip whitespace backwards
	for p.CursorCol > 0 && isWhitespace(line[p.CursorCol]) {
		p.CursorCol--
	}

	if p.CursorCol < len(line) {
		ch := line[p.CursorCol]
		if isWordChar(ch) {
			// Find start of word
			for p.CursorCol > 0 && isWordChar(line[p.CursorCol-1]) {
				p.CursorCol--
			}
		} else if !isWhitespace(ch) {
			// Find start of punctuation group
			for p.CursorCol > 0 && !isWordChar(line[p.CursorCol-1]) && !isWhitespace(line[p.CursorCol-1]) {
				p.CursorCol--
			}
		}
	}
}

// FindCharForward finds the next occurrence of ch on the current line
func (p *Pane) FindCharForward(ch rune) {
	line := p.Buffer.Lines[p.CursorRow]
	for i := p.CursorCol + 1; i < len(line); i++ {
		if line[i] == ch {
			p.CursorCol = i
			return
		}
	}
}

// FindCharBackward finds the previous occurrence of ch on the current line
func (p *Pane) FindCharBackward(ch rune) {
	line := p.Buffer.Lines[p.CursorRow]
	for i := p.CursorCol - 1; i >= 0; i-- {
		if line[i] == ch {
			p.CursorCol = i
			return
		}
	}
}

File: src/pane.c
#include "pane.h"
#include "xalloc.h"
#include "widget.h"
#include "runes.h"

#include <stdlib.h>

Pane *pane_new(Buffer *buf) {
    Pane *p = xcalloc(1, sizeof(Pane));
    p->buffer = buf;
    return p;
}

Pane *pane_new_at_line(Buffer *buf, int line) {
    Pane *p = pane_new(buf);
    if (line > 0) pane_goto_line(p, line);
    return p;
}

void pane_free(Pane *p) {
    if (!p) return;
    widget_state_free(p->widget);
    free(p);
}

bool pane_has_active_widget(const Pane *p) {
    return p && p->widget && p->widget->active;
}

void pane_goto_line(Pane *p, int line) {
    if (line < 1) line = 1;
    if (line > p->buffer->line_count) line = p->buffer->line_count;
    p->cursor_row = line - 1;
    p->cursor_col = 0;
    pane_clamp_cursor(p);
}

void pane_clamp_cursor(Pane *p) {
    if (p->cursor_row < 0) p->cursor_row = 0;
    if (p->cursor_row >= p->buffer->line_count) {
        p->cursor_row = p->buffer->line_count - 1;
    }
    if (p->cursor_row < 0) p->cursor_row = 0;
    pane_clamp_cursor_col(p);
    if (!p->selection_active) return;
    if (p->selection_start_row >= p->buffer->line_count)
        p->selection_start_row = p->buffer->line_count - 1;
    int ll = p->buffer->line_lens[p->selection_start_row];
    if (p->selection_start_col > ll) p->selection_start_col = ll;
}

void pane_clamp_cursor_col(Pane *p) {
    if (p->buffer->line_count == 0) {
        p->cursor_col = 0;
        return;
    }
    int ll = p->buffer->line_lens[p->cursor_row];
    if (p->cursor_col > ll) p->cursor_col = ll;
    if (p->cursor_col < 0) p->cursor_col = 0;
}

void pane_adjust_cursor_for_edit(Pane *p, int edit_row, int lines_delta) {
    if (lines_delta == 0) return;
    if (lines_delta > 0) {
        if (p->cursor_row >= edit_row) {
            p->cursor_row += lines_delta;
        }
    } else {
        int del_start = edit_row;
        int del_end = edit_row + (-lines_delta) - 1;
        if (p->cursor_row > del_end) {
            p->cursor_row += lines_delta;
        } else if (p->cursor_row >= del_start) {
            p->cursor_row = del_start;
        }
    }
    pane_clamp_cursor(p);
}

static int line_rows(const Pane *p, int row, int text_width) {
    int vw = buffer_get_visual_line_width(p->buffer,
        p->buffer->lines[row], p->buffer->line_lens[row]);
    return (vw == 0) ? 1 : (vw + text_width - 1) / text_width;
}

static void clamp_scroll_wrap(Pane *p, int text_width) {
    int max_wrap = line_rows(p, p->scroll_offset, text_width) - 1;
    if (p->scroll_wrap > max_wrap) p->scroll_wrap = max_wrap;
}

/* Scroll the view down by n visual rows, never past the cursor row. */
static void scroll_down_rows(Pane *p, int text_width, int n) {
    while (n > 0 && p->scroll_offset < p->cursor_row) {
        int avail = line_rows(p, p->scroll_offset, text_width) - p->scroll_wrap;
        if (avail > n) {
            p->scroll_wrap += n;
            return;
        }
        n -= avail;
        p->scroll_offset++;
        p->scroll_wrap = 0;
    }
    p->scroll_wrap += n;
}

static void scroll_up_row(Pane *p, int text_width) {
    if (p->scroll_wrap > 0) {
        p->scroll_wrap--;
    } else if (p->scroll_offset > 0) {
        p->scroll_offset--;
        p->scroll_wrap = line_rows(p, p->scroll_offset, text_width) - 1;
    }
}

void pane_adjust_scroll(Pane *p, int text_width, int screen_height) {
    if (text_width < 1) text_width = 1;

    int margin = SCROLL_MARGIN;
    if (screen_height < margin * 2 + 1) {
        margin = (screen_height - 1) / 2;
    }
    if (margin < 0) margin = 0;

    int cursor_wrap = 0;
    if (p->cursor_col > 0) {
        int vis = buffer_get_visual_column(p->buffer,
            p->buffer->lines[p->cursor_row],
            p->buffer->line_lens[p->cursor_row],
            p->cursor_col);
        cursor_wrap = vis / text_width;
    }

    /* Cursor above the view: snap the view top to the cursor row */
    if (p->cursor_row < p->scroll_offset
        || (p->cursor_row == p->scroll_offset
            && cursor_wrap < p->scroll_wrap)) {
        p->scroll_offset = p->cursor_row;
        p->scroll_wrap = cursor_wrap;
    }
    clamp_scroll_wrap(p, text_width);

    int rows_from_top = -p->scroll_wrap;
    for (int i = p->scroll_offset; i < p->cursor_row
             && i < p->buffer->line_count; i++) {
        rows_from_top += line_rows(p, i, text_width);
    }
    rows_from_top += cursor_wrap;

    /* Scroll up if cursor too close to top */
    while (rows_from_top < margin
           && (p->scroll_offset > 0 || p->scroll_wrap > 0)) {
        scroll_up_row(p, text_width);
        rows_from_top++;
    }

    /* Scroll down if cursor too close to bottom */
    int excess = rows_from_top + 1 - (screen_height - margin);
    if (excess > 0) scroll_down_rows(p, text_width, excess);
    clamp_scroll_wrap(p, text_width);
}

/* --- Selection ----------------------------------------------------------- */

void pane_start_selection(Pane *p) {
    p->selection_active = true;
    p->selection_start_row = p->cursor_row;
    p->selection_start_col = p->cursor_col;
}

void pane_clear_selection(Pane *p) {
    p->selection_active = false;
}

void pane_get_selection(const Pane *p, int *sr, int *sc, int *er, int *ec) {
    *sr = p->selection_start_row;
    *sc = p->selection_start_col;
    *er = p->cursor_row;
    *ec = p->cursor_col;
    normalize_range(sr, sc, er, ec);
}

bool pane_is_in_selection(const Pane *p, int row, int col) {
    if (!p->selection_active) return false;
    int sr, sc, er, ec;
    pane_get_selection(p, &sr, &sc, &er, &ec);
    if (row < sr || (row == sr && col < sc)) return false;
    if (row > er || (row == er && col > ec)) return false;
    return true;
}

/* --- Movement ------------------------------------------------------------ */

void pane_move_up(Pane *p) {
    if (p->cursor_row > 0) {
        p->cursor_row--;
        pane_clamp_cursor_col(p);
    }
}

void pane_move_down(Pane *p) {
    if (p->cursor_row < p->buffer->line_count - 1) {
        p->cursor_row++;
        pane_clamp_cursor_col(p);
    }
}

void pane_move_left(Pane *p) {
    if (p->cursor_col > 0) {
        p->cursor_col--;
    } else if (p->cursor_row > 0) {
        p->cursor_row--;
        p->cursor_col = p->buffer->line_lens[p->cursor_row];
    }
}

void pane_move_right(Pane *p) {
    if (p->cursor_col < p->buffer->line_lens[p->cursor_row]) {
        p->cursor_col++;
    } else if (p->cursor_row < p->buffer->line_count - 1) {
        p->cursor_row++;
        p->cursor_col = 0;
    }
}

void pane_move_up_n(Pane *p, int n) {
    for (int i = 0; i < n && p->cursor_row > 0; i++) {
        p->cursor_row--;
    }
    pane_clamp_cursor_col(p);
}

void pane_move_down_n(Pane *p, int n) {
    for (int i = 0; i < n && p->cursor_row < p->buffer->line_count - 1; i++) {
        p->cursor_row++;
    }
    pane_clamp_cursor_col(p);
}

void pane_move_left_n(Pane *p, int n) {
    for (int i = 0; i < n; i++) pane_move_left(p);
}

void pane_move_right_n(Pane *p, int n) {
    for (int i = 0; i < n; i++) pane_move_right(p);
}

void pane_move_to_line_start(Pane *p) {
    p->cursor_col = 0;
}

void pane_move_to_line_end(Pane *p) {
    p->cursor_col = p->buffer->line_lens[p->cursor_row];
}

void pane_page_down(Pane *p, int height) {
    p->cursor_row += height / 2;
    if (p->cursor_row >= p->buffer->line_count) {
        p->cursor_row = p->buffer->line_count - 1;
    }
    pane_clamp_cursor_col(p);
}

void pane_page_up(Pane *p, int height) {
    p->cursor_row -= height / 2;
    if (p->cursor_row < 0) p->cursor_row = 0;
    pane_clamp_cursor_col(p);
}

void pane_move_to_next_word(Pane *p) {
    wchar_t *line = p->buffer->lines[p->cursor_row];
    int ll = p->buffer->line_lens[p->cursor_row];

    if (p->cursor_col < ll) {
        wchar_t ch = line[p->cursor_col];
        if (is_word_char(ch)) {
            while (p->cursor_col < ll && is_word_char(line[p->cursor_col]))
                p->cursor_col++;
        } else if (!is_whitespace(ch)) {
            while (p->cursor_col < ll && !is_word_char(line[p->cursor_col])
                   && !is_whitespace(line[p->cursor_col]))
                p->cursor_col++;
        }
    }

    while (p->cursor_col < ll && is_whitespace(line[p->cursor_col]))
        p->cursor_col++;

    if (p->cursor_col >= ll && p->cursor_row < p->buffer->line_count - 1) {
        p->cursor_row++;
        p->cursor_col = 0;
        line = p->buffer->lines[p->cursor_row];
        ll = p->buffer->line_lens[p->cursor_row];
        while (p->cursor_col < ll && is_whitespace(line[p->cursor_col]))
            p->cursor_col++;
    }
}

void pane_move_to_prev_word(Pane *p) {
    if (p->cursor_col == 0 && p->cursor_row > 0) {
        p->cursor_row--;
        p->cursor_col = p->buffer->line_lens[p->cursor_row];
    }

    wchar_t *line = p->buffer->lines[p->cursor_row];
    int ll = p->buffer->line_lens[p->cursor_row];

    if (p->cursor_col > 0) p->cursor_col--;

    while (p->cursor_col > 0 && is_whitespace(line[p->cursor_col]))
        p->cursor_col--;

    if (p->cursor_col < ll) {
        wchar_t ch = line[p->cursor_col];
        if (is_word_char(ch)) {
            while (p->cursor_col > 0 && is_word_char(line[p->cursor_col - 1]))
                p->cursor_col--;
        } else if (!is_whitespace(ch)) {
            while (p->cursor_col > 0 && !is_word_char(line[p->cursor_col - 1])
                   && !is_whitespace(line[p->cursor_col - 1]))
                p->cursor_col--;
        }
    }
}

static void skip_ws_crossing_lines(Pane *p) {
    wchar_t *line = p->buffer->lines[p->cursor_row];
    int ll = p->buffer->line_lens[p->cursor_row];
    while (p->cursor_col >= ll || is_whitespace(line[p->cursor_col])) {
        if (p->cursor_col >= ll) {
            if (p->cursor_row >= p->buffer->line_count - 1) return;
            p->cursor_row++;
            p->cursor_col = 0;
            line = p->buffer->lines[p->cursor_row];
            ll = p->buffer->line_lens[p->cursor_row];
        } else p->cursor_col++;
    }
}

void pane_move_to_word_end(Pane *p) {
    int ll = p->buffer->line_lens[p->cursor_row];
    if (p->cursor_col < ll - 1) p->cursor_col++;
    else if (p->cursor_row < p->buffer->line_count - 1) {
        p->cursor_row++;
        p->cursor_col = 0;
    } else return;

    skip_ws_crossing_lines(p);
    wchar_t *line = p->buffer->lines[p->cursor_row];
    ll = p->buffer->line_lens[p->cursor_row];
    if (p->cursor_col >= ll) return;

    if (is_word_char(line[p->cursor_col])) {
        while (p->cursor_col + 1 < ll
               && is_word_char(line[p->cursor_col + 1]))
            p->cursor_col++;
    } else if (!is_whitespace(line[p->cursor_col])) {
        while (p->cursor_col + 1 < ll
               && !is_word_char(line[p->cursor_col + 1])
               && !is_whitespace(line[p->cursor_col + 1]))
            p->cursor_col++;
    }
}

void pane_find_char_forward(Pane *p, wchar_t ch) {
    wchar_t *line = p->buffer->lines[p->cursor_row];
    int ll = p->buffer->line_lens[p->cursor_row];
    for (int i = p->cursor_col + 1; i < ll; i++) {
        if (line[i] == ch) { p->cursor_col = i; return; }
    }
}

void pane_find_char_backward(Pane *p, wchar_t ch) {
    wchar_t *line = p->buffer->lines[p->cursor_row];
    for (int i = p->cursor_col - 1; i >= 0; i--) {
        if (line[i] == ch) { p->cursor_col = i; return; }
    }
}
